pkg/apiserver/profiling/service.go

```diff
@@ -145,6 +149,54 @@ func (s *Service) serviceLoop(ctx context.Context) {
 	}
 }
 
+// Migrate the legacy state TaskStateFinish for group task to TaskStateFailed|TaskStatePartialFailed|TaskStateAllSuccess
+func (s *Service) migrateLegacyState(ctx context.Context) {
+	var lastMinID uint
+	var groupTasks []TaskGroupModel
+	var tasks []TaskModel
+	for {
+		// Step 1: find out group tasks whose state is TaskStateFinish
+		query := s.params.LocalStore.Where("state = ?", TaskStateFinish).Order("id DESC").Limit(100)
+		if lastMinID > 0 {
+			// Add this where condition to avoid the loop can't escape when failed to modify the TaskStateFinish to other states for some group tasks
+			query = query.Where("id < ?", lastMinID)
+		}
+		err := query.Find(&groupTasks).Error
+		if err != nil || len(groupTasks) == 0 {
+			break
+		}
+		lastMinID = groupTasks[len(groupTasks)-1].ID
+
+		for _, groupTask := range groupTasks {
+			// Step 2: find out all child tasks of the group task
+			err = s.params.LocalStore.Where("task_group_id = ?", groupTask.ID).Find(&tasks).Error
+			if err != nil {
+				continue
+			}
+			if len(tasks) == 0 {
+				s.params.LocalStore.Delete(groupTask)
+				continue
+			}
+
+			// Step 3: fix the group task state
+			successTasks := 0
+			for _, t := range tasks {
+				if t.State == TaskStateFinish {
+					successTasks++
+				}
+			}
+			if successTasks == 0 {
+				groupTask.State = TaskStateFailed
+			} else if successTasks < len(tasks) {
+				groupTask.State = TaskStatePartialFailed
+			} else {
+				groupTask.State = TaskStateAllSuccess
+			}
+			s.params.LocalStore.Save(groupTask)
```

Comment:
This way migration can't handle the situation when down-grade TiDB version, for example from TiDB 5.3.0 back to TiDB 5.1.2.

I have a new idea.